core/chain-configs/src/config_store.rs

```diff
@@ -0,0 +1,111 @@
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use crate::ChainConfig;
+use crate::config::ChainConfigPatch;
+use crate::genesis_config::GenesisConfig;
+use std::fs::File;
+use std::io::Read;
+use serde_json;
+use std::env;
+use std::ops::Bound;
+use near_primitives::types::ProtocolVersion;
+
+pub const RESOURCES_DIR: &str = "core/chain-configs/res/";
+pub const CONFIG_PATCH_LIST_FILENAME: &str = "config_patch_list.json";
+
+/// Stores protocol versions when new config patches should be applied.
+/// This list is used to load files with config patches.
+#[derive(Debug, serde::Deserialize)]
+pub struct ConfigPatchList {
+    protocol_version_patches: Vec<ProtocolVersion>,
+}
+
+/// Stores chain config for each protocol version where it was updated.
+#[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize)]
+pub struct ChainConfigStore {
+    /// Initial chain config at genesis block.
+    pub initial_chain_config: Arc<ChainConfig>,
+    /// Maps protocol version to the config.
+    store: BTreeMap<ProtocolVersion, Arc<ChainConfig>>,
+}
+
+impl ChainConfigStore {
+    /// Constructs a new store. Initial chain config is read from provided genesis_config.
+    /// Stored values are read from config path list file and specific patches are read from the
+    /// files that are named after the protocol version when they should be applied.
+    pub fn new(genesis_config: GenesisConfig) -> Self {
+        let initial_chain_config = Arc::new(ChainConfig::new(genesis_config.clone()));
+        let mut store = BTreeMap::new();
+        store.insert(0, initial_chain_config.clone());
+        Self::populate_config_store(*initial_chain_config, &mut store);
+        Self { initial_chain_config, store }
+    }
+
+    fn populate_config_store(
+        initial_chain_config: ChainConfig,
+        store: &mut BTreeMap<ProtocolVersion, Arc<ChainConfig>>,
+    ) {
+        let config_patch_list = Self::get_config_patch_list();
+        let mut curr_config = initial_chain_config.clone();
+
+        for protocol_veriosn in config_patch_list.protocol_version_patches {
+            let chain_config_patch = Self::load_chain_config_patch(protocol_veriosn);
+            curr_config = curr_config.apply_patch(&chain_config_patch);
+
+            store.insert(protocol_veriosn, Arc::new(curr_config.clone()));
+        }
+    }
+
+    fn load_chain_config_patch(protocol_version: ProtocolVersion) -> ChainConfigPatch {
+        let current_dir = env::current_dir().expect("Failed to get the current directory");
+        let path = current_dir.join(RESOURCES_DIR).join(protocol_version.to_string() + ".json");
```

Comment:
That's too complicated. I'd expect the overrides config to be in `~/.near/`, next to `config.json`.
I'm also confused why you need protocol version in the name of the file.
Wouldn't it be simpler to have a json object that contains a map of protocol versions to the overrides?

Reply:
Sure, I can place the file next to `config.json` and skip it if it is not provided. I do not need the protocol version in the name of the file, I just used it to resemble the runtime config style. I can put it in a format where there is a map from the protocol version to the overrides. I will implement it in the proposed way since I do not find a great difference between the proposed approaches.

Can you please take a look at the rest of the code so that I can proceed with implementing the actual PR?